support/agent/miner_agent.go

```diff
@@ -232,25 +240,59 @@ func (ma *MinerAgent) createPreCommit(currentEpoch abi.ChainEpoch) message {
 		SealRandEpoch: currentEpoch - 1,
 		Expiration:    expiration,
 	}
-	ma.pendingDeals = nil
+
+	// upgrade sector if upgrades are on, this sector has deals, and we have a cc sector
+	if ma.Config.UpgradeSectors && len(dealIds) > 0 && len(ma.ccSectors) > 0 {
+		var upgradeNumber uint64
+		upgradeNumber, ma.ccSectors = PopRandom(ma.ccSectors, ma.rnd)
+
+		// prevent sim from attempting to upgrade to sector with shorter duration
+		sinfo, err := ma.sectorInfo(s, upgradeNumber)
+		if err != nil {
+			return message{}, err
+		}
+		if sinfo.Expiration > expiration {
+			params.Expiration = sinfo.Expiration
+		}
+
+		dlInfo, pIdx, err := ma.dlInfoForSector(s, upgradeNumber)
+		if err != nil {
+			return message{}, err
+		}
+
+		params.ReplaceCapacity = true
+		params.ReplaceSectorNumber = abi.SectorNumber(upgradeNumber)
+		params.ReplaceSectorDeadline = dlInfo.Index
+		params.ReplaceSectorPartition = pIdx
+		ma.UpgradedSectors++
```

Comment:
Because this is updated at precommit the name is a bit confusing.  Either this should be updated on prove or the name changed to reflect that not all of these upgrades are successful.

Reply:
I delayed this stat update till the sector is registered (after CSPV).